File: networking/src/main/java/work/fking/pangya/networking/TransportUtils.java
package work.fking.pangya.networking;

import io.netty.channel.Channel;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.ServerChannel;
import io.netty.channel.epoll.Epoll;
import io.netty.channel.epoll.EpollEventLoopGroup;
import io.netty.channel.epoll.EpollServerSocketChannel;
import io.netty.channel.epoll.EpollSocketChannel;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.channel.socket.nio.NioSocketChannel;

public final class TransportUtils {

    private TransportUtils() {
        throw new UnsupportedOperationException("Utility class");
    }

    public static Class<? extends Channel> selectBestChannelAvailable() {
        return Epoll.isAvailable() ? EpollSocketChannel.class : NioSocketChannel.class;
    }

    public static Class<? extends ServerChannel> selectBestServerChannelAvailable() {
        return Epoll.isAvailable() ? EpollServerSocketChannel.class : NioServerSocketChannel.class;
    }

    public static EventLoopGroup selectBestEventLoopAvailable() {
        return selectBestEventLoopAvailable(0);
    }

    public static EventLoopGroup selectBestEventLoopAvailable(int threads) {
        return Epoll.isAvailable() ? new EpollEventLoopGroup(threads) : new NioEventLoopGroup(threads);
    }
}

File: networking/src/main/kotlin/work/fking/pangya/networking/TransportUtils.kt
package work.fking.pangya.networking

import io.netty.channel.EventLoopGroup
import io.netty.channel.ServerChannel
import io.netty.channel.epoll.Epoll
import io.netty.channel.epoll.EpollEventLoopGroup
import io.netty.channel.epoll.EpollServerSocketChannel
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioServerSocketChannel

fun selectBestServerChannelAvailable(): Class<out ServerChannel?> {
    return if (Epoll.isAvailable()) EpollServerSocketChannel::class.java else NioServerSocketChannel::class.java
}

fun selectBestEventLoopAvailable(threads: Int = 1): EventLoopGroup {
    return if (Epoll.isAvailable()) EpollEventLoopGroup(threads) else NioEventLoopGroup(threads)
}
